### PR TITLE
Fix/salary average and cities names in country

### PR DESCRIPTION
**Pull request by HaidyAbuGom3a**

### CODE REVIEW

**Review by flutterdevtaha — APPROVED** [review]

**Comment by flutterdevtaha** [comment]
[Enhancement on Task 3](https://trello.com/c/SiAmzSFy/32-enhancement-on-task-3)

File: src/main/kotlin/interactor/GetSalaryAverageAndCitiesNamesInCountryInteractor.kt [hunk 1]
```diff
@@ -1,28 +1,23 @@
...
-        else
-            throw Exception("Invalid country name")
-    }
-
-    private fun checkIfCountryIsValid(countryName: String):Boolean{
-        return countryName.lowercase() in dataSource.getAllCitiesData()
-            .map{it.country.lowercase()}
-    }
+        return dataSource
+            .getAllCitiesData()
```
Comment (ma28alhams) [inline src/main/kotlin/interactor/GetSalaryAverageAndCitiesNamesInCountryInteractor.kt:12]: To make sure about the returned data from the data source, check if it is an empty list or not, if it is then throw IllegalStateException("Something went wrong")
```suggestion
            .getAllCitiesData()
            .let{ it.ifEmpty{ IllegalStateException("Something went wrong") } }
```

File: src/main/kotlin/interactor/GetSalaryAverageAndCitiesNamesInCountryInteractor.kt [hunk 2]
```diff
@@ -1,28 +1,23 @@
...
+            .filter{
+                    it.averageMonthlyNetSalaryAfterTax.isNotNull() &&
+                    it.dataQuality &&
+                    it.country.lowercase() == countryName.lowercase()
+            }
+            .map{Pair(it.cityName,it.averageMonthlyNetSalaryAfterTax!!)}
+            .takeIf { it.isNotEmpty() } ?: throw InvalidCountryNameException("Invalid country name")
```
Comment (ma28alhams) [inline src/main/kotlin/interactor/GetSalaryAverageAndCitiesNamesInCountryInteractor.kt:19]: Why should we go through the entire code, when the country name is wrong?

File: src/main/kotlin/interactor/GetSalaryAverageAndCitiesNamesInCountryInteractor.kt [hunk 3]
```diff
@@ -1,28 +1,23 @@
...
 
-    private fun excludeNullSalariesAndLowQualityData(city: CityEntity): Boolean {
-        return city.averageMonthlyNetSalaryAfterTax != null && city.dataQuality
```
Comment (ma28alhams) [inline src/main/kotlin/interactor/GetSalaryAverageAndCitiesNamesInCountryInteractor.kt:25]: Why have you deleted these lines??

File: src/test/kotlin/interactor/GetSalaryAverageAndCitiesNamesInCountryInteractorTest.kt [hunk 4]
```diff
@@ -23,51 +23,78 @@ internal class GetSalaryAverageAndCitiesNamesInCountryInteractorTest {
...
         //then
-        assertEquals(listOf(Pair("Havana",35.75f)),getSalaryAverageAndCitiesNamesInCountry)
+        assertEquals(expected,actual)
     }
 
     @Test
     fun `should throw exception when the country name is wrong`() {
         //given wrong country name
-        val country = fakeDataSource.getAllCitiesData()[0].country.substring(0,2)
+        val country = "lol"
```
Comment (ma28alhams) [inline src/test/kotlin/interactor/GetSalaryAverageAndCitiesNamesInCountryInteractorTest.kt:59]: "lol" i like that name :1st_place_medal:

File: src/test/kotlin/interactor/GetSalaryAverageAndCitiesNamesInCountryInteractorTest.kt [hunk 5]
```diff
@@ -23,51 +23,78 @@ internal class GetSalaryAverageAndCitiesNamesInCountryInteractorTest {
...
-        val getSalaryAverageAndCitiesNamesInCountry= Executable{interactor.execute(country)}
+        val actual = Executable { interactor.execute(country) }
+        val expected = Exception::class.java
         //then
-        assertThrows(Exception::class.java,getSalaryAverageAndCitiesNamesInCountry)
+        assertThrows(expected, actual)
     }
 
+    @Test
+    fun `should throw exception when the country name is correct but excluded`(){
```
Comment (ma28alhams) [inline src/test/kotlin/interactor/GetSalaryAverageAndCitiesNamesInCountryInteractorTest.kt:79]: should throw an empty list if the filtered list is empty, 
but if the returned data from the data source is empty, then should throw IllegalStateException

File: src/test/kotlin/interactor/GetSalaryAverageAndCitiesNamesInCountryInteractorTest.kt [hunk 6]
```diff
@@ -23,51 +23,78 @@ internal class GetSalaryAverageAndCitiesNamesInCountryInteractorTest {
...
+        //when getting a list of pair contains the city name and average salary of country with wrong name
+        val actual = Executable { interactor.execute(country) }
+        val expected = Exception::class.java
+        //then
+        assertThrows(expected, actual)
+
+    }
+
+    @Test
+    fun `should throw exception when the country name is correct but not in the data`(){
```
Comment (ma28alhams) [inline src/test/kotlin/interactor/GetSalaryAverageAndCitiesNamesInCountryInteractorTest.kt:91]: Should return empty list for that case

**Review by ma28alhams — CHANGES_REQUESTED** [review]
Good work, but some changes are required :100:

File: src/main/kotlin/interactor/GetSalaryAverageAndCitiesNamesInCountryInteractor.kt (hunk 2 above)
Comment (HaidyAbuGom3a) [reply src/main/kotlin/interactor/GetSalaryAverageAndCitiesNamesInCountryInteractor.kt:19]: in the code review bareq said that the checkIfCountryIsValid function is repeating in the code as we will access the data source and use getAllCititesData twice so we can use takeIf and thus if the country name is valid but excluded with null data or the country name is not valid we will throw an exception as the result of takeIf would be null in these cases

File: src/main/kotlin/interactor/GetSalaryAverageAndCitiesNamesInCountryInteractor.kt (hunk 3 above)
Comment (HaidyAbuGom3a) [reply src/main/kotlin/interactor/GetSalaryAverageAndCitiesNamesInCountryInteractor.kt:25]: i used them in the filter instead
for more details check the code review with lemon cake team in this link https://youtu.be/AHevF7MlP6E
you don't need to search for the task as I provided it in the link

File: src/test/kotlin/interactor/GetSalaryAverageAndCitiesNamesInCountryInteractorTest.kt (hunk 5 above)
Comment (HaidyAbuGom3a) [reply src/test/kotlin/interactor/GetSalaryAverageAndCitiesNamesInCountryInteractorTest.kt:79]: ok got it !

File: src/main/kotlin/interactor/GetSalaryAverageAndCitiesNamesInCountryInteractor.kt (hunk 1 above)
Comment (flutterdevtaha) [reply src/main/kotlin/interactor/GetSalaryAverageAndCitiesNamesInCountryInteractor.kt:12]: you say, that we should not return exception if the list is empty. why here you suggested the oppisite?

**Comment by HaidyAbuGom3a** [comment]
all changes required are done.
please review my code

**Pull request merged**